client/request_test.go

```diff
@@ -0,0 +1,72 @@
+package client
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"testing"
+
+	"golang.org/x/net/context"
+)
+
+// TestSetHostHeader should set fake host for local communications, set real host
+// for normal communications.
+func TestSetHostHeader(t *testing.T) {
+	testURL := "/test"
+	testCases := []struct {
+		host            string
+		expectedHost    string
+		expectedURLHost string
+	}{
+		{
+			"unix:///var/run/docker.sock",
+			"",
+			"docker",
+		},
+		{
+			"npipe:////./pipe/docker_engine",
+			"",
+			"docker",
+		},
+		{
+			"tcp://0.0.0.0:4243",
+			"",
```

Comment:
@duglin, I think the main reason is that you added `req.Host = host` in request.go, there is nothing to do with the `NewClient`.

The `req.Host` is initialized in `http.NewRequest` before, and mostly it will be set to "" except the bug I mentioned [above](https://github.com/docker/engine-api/pull/190/files#r58492191).

Now you manually added `req.Host = host`, it will be sure to be the same with `req.URL.Host`.